api/http_verbs_with_json.go

```diff
@@ -0,0 +1,61 @@
+package api
+
+import (
+	"errors"
+
+	"github.com/go-resty/resty/v2"
+)
+
+func (self *ApiClient) postJSON(path string, request interface{}, response interface{}) error {
+	return self.executeJSONInJSONOut(resty.MethodPost, path, request, response)
+}
+
+func (self *ApiClient) putJSON(path string, request interface{}, response interface{}) error {
+	return self.executeJSONInJSONOut(resty.MethodPut, path, request, response)
+}
+
+func (self *ApiClient) executeJSONInJSONOut(method string, path string, request interface{}, response interface{}) error {
+	self.normalizeEndpoint()
+	result, err := self.client.R().
+		SetBasicAuth(self.ApiKey, self.ApiSecret).
+		SetBody(request).
+		SetResult(response).
+		Execute(method, self.Endpoint+path)
+	if err != nil {
+		return err
+	}
+	if result.StatusCode() < 200 || result.StatusCode() > 299 {
+		return errors.New(result.Status() + ": " + string(result.Body()))
+	}
+	return nil
+}
+
+func (self *ApiClient) getJSON(path string, params map[string]string, response interface{}) error {
+	self.normalizeEndpoint()
+	result, err := self.client.R().
+		SetBasicAuth(self.ApiKey, self.ApiSecret).
+		SetQueryParams(params).
+		SetResult(response).
+		Get(self.Endpoint + path)
+	if err != nil {
+		return err
+	}
+	if result.StatusCode() < 200 || result.StatusCode() > 299 {
+		return errors.New(result.Status() + ": " + string(result.Body()))
+	}
+	return nil
```

Comment:
These lines of code seems to repeat for every verb.  
Let's promote code reuse by extracting two functions. One that would:
1. Normalizes the endpoint (do we really have to do that? doesn't the internal client offer a way to do that?)
2. Sets the basic auth
2. Sets host url

And another function that would handle all the errors and error code.  

Put together, this file should ultimately consist of 4 key verb functions, something tin the lines of:    

```go
func (api *ApiClient) get(path string, params QueryParams, response interface{}) error {
  // api.getClient() is a function that returns
  // .R().SetBasicAuth(self.ApiKey, self.ApiSecret).SetHostURL(self.Endpoint)
  result, err = api.getClient()
        .SetQueryParams(params)
        .SetResult(response)
        .Get(path);
  return httpResult(result, err); // a function that handles non 2xx status codes/err returned 
}
```